### PR TITLE
Slightly refactor PF endcap cluster calibration

### PR DESCRIPTION
**Pull request by guitargeek**
#### PR description:

In the PFEgammaAlgo, there is still this comment about code "stolen" from the PFECALSuperClusterAlgo:
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc#L1774

I was thinking of ways to avoid this duplication, but it's not quite trivial because the code diverged a bit. However, a bit part of the duplicate code serves to compute cluster energy corrections, so I could move this into the PFEnergyCalibration class which is used by both algos. I think this is a good step to avoid this code duplication, but I don't want to go further for now to not risk accidentally changing the logic. What do you think?

#### PR validation:

CMSSW compiles and matrix tests pass.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

No backport intended.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by silviodonato** [comment]
Do you have any comments? @perrotta @slava77

File: RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc [hunk 1]
```diff
@@ -161,6 +161,59 @@ void PFEnergyCalibration::initializeCalibrationFunctions() {
...
+PFEnergyCalibration::CalibratedEndcapPFClusterEnergies PFEnergyCalibration::calibrateEndcapClusterEnergies(
+    reco::PFCluster const& eeCluster,
+    std::vector<reco::PFCluster const*> const& psClusterPointers,
+    ESChannelStatus const& channelStatus,
+    bool applyCrackCorrections) const {
+  double ps1_energy_sum = 0.;
+  double ps2_energy_sum = 0.;
+  double ePS1 = 0.;
+  double ePS2 = 0.;
+  int condP1 = 1;
```
Comment (perrotta) [inline RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc]: Maybe if you make these two some "bool" variable, everything gets more intuitive

File: RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc [hunk 2]
```diff
@@ -161,6 +161,59 @@ void PFEnergyCalibration::initializeCalibrationFunctions() {
...
+    switch (psclus->layer()) {
+      case PFLayer::PS1:
+        ps1_energy_sum += psclus->energy();
+        for (auto const& recH : recH_Frac) {
+          ESDetId strip1 = recH.recHitRef()->detId();
+          if (strip1 != ESDetId(0)) {
+            //getStatusCode() == 0 => active channel
+            // apply correction if all recHits are dead
+            if (channelStatus.getMap().find(strip1)->getStatusCode() == 0)
+              condP1 = 0;
```
Comment (perrotta) [inline RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc]: you can break here

File: RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc [hunk 3]
```diff
@@ -161,6 +161,59 @@ void PFEnergyCalibration::initializeCalibrationFunctions() {
...
+          }
+        }
+        break;
+      case PFLayer::PS2:
+        ps2_energy_sum += psclus->energy();
+        for (auto const& recH : recH_Frac) {
+          ESDetId strip2 = recH.recHitRef()->detId();
+          if (strip2 != ESDetId(0)) {
+            if (channelStatus.getMap().find(strip2)->getStatusCode() == 0)
+              condP2 = 0;
```
Comment (perrotta) [inline RecoParticleFlow/PFClusterTools/src/PFEnergyCalibration.cc]: you can break here

File: RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc [hunk 4]
```diff
@@ -316,62 +316,25 @@ void PFECALSuperClusterAlgo::buildSuperCluster(CalibClusterPtr& seed, CalibClust
...
   // calculate necessary parameters and build the SC
-  double posX(0), posY(0), posZ(0), corrSCEnergy(0), corrPS1Energy(0), corrPS2Energy(0), ePS1(0), ePS2(0),
-      energyweight(0), energyweighttot(0);
-  std::vector<double> ps1_energies, ps2_energies;
-  int condP1(1), condP2(1);
+  double posX(0), posY(0), posZ(0), corrSCEnergy(0), corrPS1Energy(0), corrPS2Energy(0), energyweight(0),
+      energyweighttot(0);
   for (auto& clus : clustered) {
-    ePS1 = ePS2 = 0;
+    double ePS1 = 0.0;
```
Comment (perrotta) [inline RecoEcal/EgammaClusterAlgos/src/PFECALSuperClusterAlgo.cc]: (just for symmetry) you could write "0" for both them...

**Comment by guitargeek** [comment]
Hi @perrotta, I addressed your comments and also got rid of an "temporaty ugly fix" in the PFEnergyCalibration class. Just reading the code of the PFCluster and CaloCluster classes, it seems equivalent to just use the `eta()` and `phi()` member functions without having to copy the cluster, but I guess the bot comparisons will tell if this is really correct.

**Comment by perrotta** [comment]
please test

**Comment by guitargeek** [comment]
Hi @perrotta, I made a mistake in my previous commit. Now I fixed that commit and also made sure the local tests pass. So the bot should succeed now!

**Comment by perrotta** [comment]
please test

**Comment by guitargeek** [comment]
Hi @perrotta, inspired by your previous comments that suggested to use `bool` and not `double` to encode whether the PS planes should be used for the corrections or not, I propagated this change now through all the functions of the class. That makes it more understandable, as using the same variables to declare the PS plane status and store the output was confusing.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
> * Reco comparison results: 1530 differences found in the comparisons

The latest commit wasn't that innocuos... We must check where those differences come from.
Perhaps, factorizing the updates whenever possible could be a better option: it would at least be easier to review

**Comment by guitargeek** [comment]
Yes that's right, it would have been nice if it worked but I tried to do too much at once at this PR. I will amend the last commit and only keep the changes that are hopefully really innocent, and then maybe we can do the rest at some other time.

**Comment by guitargeek** [comment]
Hi @perrotta, now it's ready to be tested again. I only kept the change with the other overloaded version of the `energyEm` function in the `ConvBremPFTrackFinder` function, so that the interface of `PFEnergyCalibration` needs fewer public functions. This should not change the result.

**Comment by perrotta** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
+1
- It allows some refactorization of the PF energy calibration 
- Jenkins tests pass and show no difference with the current reco outputs, as it should

**Comment by silviodonato** [comment]
+1

**Comment by slava77** [comment]
[I'm not sure if this is the PR with the problem, but the error messages appear to point to this]

@smuzaffar @mrodozov 
please remind me how to test on PPC interactively.
I suspect that the problem came from this PR, based on 140.52 working in CMSSW_11_1_X_2020-03-08-2300 but not in CMSSW_11_1_X_2020-03-09-2300.
In the latter we have
```
An exception of category 'StdException' occurred while
   [0] Processing  stream begin Run run: 152698 stream: 0
   [1] Calling method for module PFProducer/'particleFlowTmp'
   [2] Using EventSetup component PoolDBESSource/'GlobalTag' to make data PerformancePayload/'' in record PFCalibrationRcd
Exception Message:
A std::exception was thrown.
getrandom
```

https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_ppc64le_gcc820/CMSSW_11_1_X_2020-03-08-2300/pyRelValMatrixLogs/run/140.52_RunHI2010+RunHI2010+RECOHID10+RECOHIR10D11+HARVESTDHI/step2_RunHI2010+RunHI2010+RECOHID10+RECOHIR10D11+HARVESTDHI.log
vs
https://cmssdt.cern.ch/SDT/cgi-bin/buildlogs/slc7_ppc64le_gcc820/CMSSW_11_1_X_2020-03-09-2300/pyRelValMatrixLogs/run/140.52_RunHI2010+RunHI2010+RECOHID10+RECOHIR10D11+HARVESTDHI/step2_RunHI2010+RunHI2010+RECOHID10+RECOHIR10D11+HARVESTDHI.log

**Comment by slava77** [comment]
ping

https://github.com/cms-sw/cmssw/pull/28960#issuecomment-597599381 
> @smuzaffar @mrodozov
please remind me how to test on PPC interactively.

**Comment by smuzaffar** [comment]
@slava77 send me your ssh public key

**Comment by slava77** [comment]
I managed to get to the error in PerformancePayload.
It is apparently unrelated to this PR.
The same IB where the problem appeared apparently also has https://github.com/cms-sw/cmsdist/pull/5573
the gdb stack trace points to a problem in boost and simply merging this PR on top of 08-2300 IB does not introduce any problems.

**Pull request merged**